apps/src/gamelab/blocks.js

```diff
@@ -551,14 +593,143 @@ export default {
+      init: function () {
+        var showParamEditIcon = !Blockly.disableParamEditing && !Blockly.useModalFunctionEditor;
+
+        this.setHSV(136, 0.84, 0.80);
+        var name = Blockly.Procedures.findLegalName(
+            Blockly.Msg.PROCEDURES_DEFNORETURN_PROCEDURE, this);
+        this.appendDummyInput()
+            .appendTitle(showParamEditIcon ? '' : ' ')
+            .appendTitle(
+                new Blockly.FieldTextInput(name, Blockly.Procedures.rename),
+                'NAME')
+            .appendTitle('', 'PARAMS');
+        this.appendStatementInput('STACK')
+            .appendTitle(Blockly.Msg.PROCEDURES_DEFNORETURN_DO);
+        if (showParamEditIcon) {
+          this.setMutator(new Blockly.Mutator(['procedures_mutatorarg']));
+        }
+        // Only want to have the backdrop in the mainBlockSpace. We don't want it in
+        // the toolbox or in the feedback dialog (which is readonly).
+        this.setFramed( this.blockSpace === Blockly.mainBlockSpace &&
+            !this.blockSpace.isReadOnly());
+        this.parameterNames_ = ['this sprite'];
+        this.parameterTypes_ = [Blockly.BlockValueType.SPRITE];
+      },
+      updateParams_: Blockly.Blocks.procedures_defnoreturn.updateParams_,
+      mutationToDom: Blockly.Blocks.procedures_defnoreturn.mutationToDom,
+      domToMutation: Blockly.Blocks.procedures_defnoreturn.domToMutation,
+      updateParamsFromArrays: Blockly.Blocks.procedures_defnoreturn.updateParamsFromArrays,
+      updateCallerParams_: Blockly.Blocks.procedures_defnoreturn.updateCallerParams_,
+      dispose: Blockly.Blocks.procedures_defnoreturn.dispose,
+      getProcedureInfo: function () {
+        return {
+          name: this.getTitleValue('NAME'),
+          parameterNames: this.parameterNames_,
+          parameterIDs: this.paramIds_,
+          parameterTypes: this.parameterTypes_,
+          type: this.type,
+          callType: this.callType_
+        };
+      },
+      getVars: function (category) {
+        return {
+          Behavior: [this.getTitleValue('NAME')],
+        };
+      },
+      renameVar: Blockly.Blocks.procedures_defnoreturn.renameVar,
+      removeVar: Blockly.Blocks.procedures_defnoreturn.removeVar,
+      userCreated: false,
+      shouldBeGrayedOut: function () {
+        return false;
+      },
+      callType_: 'gamelab_behavior_get'
+    };
+
+    generator.behavior_definition = generator.procedures_defnoreturn;
+
+    Blockly.Procedures.DEFINITION_BLOCK_TYPES.push('behavior_definition');
+    Blockly.Variables.getters[Blockly.BlockValueType.BEHAVIOR] = 'gamelab_behavior_get';
+    Blockly.Flyout.configure(Blockly.BlockValueType.BEHAVIOR, {
+      initialize(flyout, cursor) {
+        if (Blockly.functionEditor && !Blockly.functionEditor.isOpen()) {
+          flyout.addButtonToFlyout_(cursor, 'Create a Behavior',
+            behaviorEditor.openWithNewFunction.bind(behaviorEditor));
+        }
+      },
+      addDefaultVar: false,
+    });
```

Comment:
I don't love how much manual work is required here to get this to work, particularly given how much of it is just copying stuff from `procedures_defnoreturn`. Can some of this be pushed into Blockly and generalized? (possibly in a separate PR)

Reply:
I added a helper in https://github.com/code-dot-org/blockly/pull/107 and used it in https://github.com/code-dot-org/code-dot-org/pull/22307/commits/1a2037d411184c3ebc4ea7d5438426c5fc4e8aa5, cutting out quite a lot of the extra junk here.